src/Selda/PG/Aff.purs

```diff
@@ -35,34 +43,90 @@ type B = BackendPGClass
+      (RecordToArrayForeign BackendPGClass)
+      (Array Foreign)
+      { | r }
+      (Array Foreign)
   ⇒ Connection → Table t → Array { | r } → Aff (Either PGError Unit)
-insert_ conn t r = runSelda conn $ Selda.PG.insert_ t r
+insert_ conn t r =
+  runSelda conn $ Selda.PG.insert_ t r
+
+insert_'
+  ∷ ∀ t r
+  . GenericInsert BackendPGClass PGSelda t r
+  ⇒ Connection
+  → Table t
+  → ({ | r } → Array Foreign)
+  → Array { | r }
+  → Aff (Either PGError Unit)
+insert_' conn t decodeRow r =
+  runSelda conn $ Selda.PG.insert_' t decodeRow r
 
 insert1_
   ∷ ∀ r t
   . GenericInsert BackendPGClass PGSelda t r
-  ⇒ Connection → Table t → { | r } → Aff (Either PGError Unit)
+  ⇒ HFoldl
+      (RecordToArrayForeign BackendPGClass)
+      (Array Foreign)
+      { | r }
+      (Array Foreign)
+  ⇒ Connection
+  → Table t
+  → { | r }
+  → Aff (Either PGError Unit)
 insert1_ conn t r = runSelda conn $ Selda.PG.insert1_ t r
 
-insert
-  ∷ ∀ r t tr
-  . InsertRecordIntoTableReturning r t tr
-  ⇒ Connection → Table t → Array { | r } → Aff (Either PGError (Array { | tr }))
-insert conn t r = runSelda conn $ Selda.PG.insert t r
+insert1_'
+  ∷ ∀ r t
+  . GenericInsert BackendPGClass PGSelda t r
+  ⇒ Connection
+  → Table t
+  → ({ | r } → Array Foreign)
+  → { | r }
+  → Aff (Either PGError Unit)
+insert1_' conn t encode r = runSelda conn $ Selda.PG.insert1_' t encode r
+
+-- insert
+--   ∷ ∀ r t ret rTuple tr trTuple s
+--   . InsertRecordIntoTableReturning r t ret
+--   ⇒ ToSQLRow rTuple
+--   ⇒ FromSQLRow trTuple
+--   ⇒ HFoldl RecordToTuple Unit { | r } rTuple
+--   ⇒ ColsToPGHandler s tr trTuple ret
+--   ⇒ TableToColsWithoutAlias s t tr
+--   ⇒ Connection → Table t → Array { | r } → Aff (Either PGError (Array { | ret }))
+-- insert conn t r = runSelda conn $ Selda.PG.insert t r
 
-insert1
-  ∷ ∀ r t tr
-  . InsertRecordIntoTableReturning r t tr
-  ⇒ Connection → Table t → { | r } → Aff (Either PGError { | tr })
-insert1 conn t r = runSelda conn $ Selda.PG.insert1 t r
+-- insert1
+--   ∷ ∀ r t ret rTuple tr trTuple s
+--   . InsertRecordIntoTableReturning r t ret
+--   ⇒ ToSQLRow rTuple
+--   ⇒ FromSQLRow trTuple
+--   ⇒ HFoldl RecordToTuple Unit { | r } rTuple
+--   ⇒ ColsToPGHandler s tr trTuple ret
+--   ⇒ TableToColsWithoutAlias s t tr
+--   ⇒ Connection → Table t → { | r } → Aff (Either PGError { | ret })
+-- insert1 conn t r = runSelda conn $ Selda.PG.insert1 t r
 
 query
   ∷ ∀ o i tup
   . ColsToPGHandler B i tup o
   ⇒ GetCols i
   ⇒ FromSQLRow tup
-  ⇒ Connection → FullQuery B (Record i) → Aff (Either PGError (Array { | o }))
-query conn q = runSelda conn $ Selda.PG.query q
+  ⇒ Connection
+  → FullQuery B (Record i)
+  → Aff (Either PGError (Array { | o }))
+query conn = runSelda conn <<< Selda.PG.query
+
+query'
+  ∷ ∀ o i
+  . GetCols i
+  ⇒ Connection
+  → FullQuery B { | i }
+  → (Array Foreign -> Either String { | o })
+  → Aff (Either PGError (Array { | o }))
+query' conn q decodeRow = runSelda conn $ Selda.PG.query' q decodeRow
```

Comment:
Eta-reducing these produces compiler errors.

Reply:
what do you mean? there's `$` so the term cannot be eta-reduced

Reply:
I think I was thinking `query' conn = runSelda conn <<< Selda.PG.query'`. It's been so long since I worked on this, I'm not sure anymore.